crates/tinymist-lint/src/lib.rs

```diff
@@ -31,28 +34,74 @@ pub struct LintInfo {
 }
 
 /// Performs linting check on file and returns a vector of diagnostics.
-pub fn lint_file(world: &LspWorld, expr: &ExprInfo, ti: Arc<TypeInfo>) -> LintInfo {
-    let diagnostics = Linter::new(world, ti).lint(expr.source.root());
+pub fn lint_file(
+    world: &LspWorld,
+    ei: &ExprInfo,
+    ti: Arc<TypeInfo>,
+    known_issues: &KnownLintIssues,
+) -> LintInfo {
+    let diagnostics = Linter::new(world, ei.clone(), ti, known_issues).lint(ei.source.root());
     LintInfo {
-        revision: expr.revision,
-        fid: expr.fid,
+        revision: ei.revision,
+        fid: ei.fid,
         diagnostics,
     }
 }
 
-struct Linter<'w> {
+/// Information about issues the linter checks for that will already be reported to the user via
+/// other means (such as compiler diagnostics), to avoid duplicating warnings.
+pub struct KnownLintIssues {
```

Comment:
We could consider computing `KnownLintIssues` lazily (i.e., only compute `unknown_vars` from the compiler diagnostics when we check for a duplicate warning.) It's kind of annoying though because that means we have to store the compiler diagnostics in `KnownLintIssues`, which means either an allocation from collecting into a Vec or a generic type parameter (since the iterator type is not concrete.)

IMO it's not worth the complexity especially since usually there are not many compiler diagnostics, so the current code computes eagerly in the `from_compiler_diagnostics` constructor.